sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java

```diff
@@ -235,41 +264,58 @@ public void process(@DoFn.Element DataChangeRecord record, OutputReceiver<Row> r
                 .map(nonNullValues -> getGson().fromJson(nonNullValues, Map.class))
                 .orElseGet(Collections::emptyMap);
 
-        for (Map.Entry<String, String> valueEntry : newValues.entrySet()) {
-          if (valueEntry.getValue() == null) {
-            continue;
+        try {
+          for (Map.Entry<String, String> valueEntry : newValues.entrySet()) {
+            if (valueEntry.getValue() == null) {
+              continue;
+            }
+            // TODO(pabloem): Understand why SpannerSchema has field names in lowercase...
+            rowBuilder =
+                rowBuilder.withFieldValue(
+                    valueEntry.getKey().toLowerCase(),
+                    stringToParsedValue(
+                        internalRowSchema.getField(valueEntry.getKey().toLowerCase()).getType(),
+                        valueEntry.getValue()));
           }
-          // TODO(pabloem): Understand why SpannerSchema has field names in lowercase...
-          rowBuilder =
-              rowBuilder.withFieldValue(
-                  valueEntry.getKey().toLowerCase(),
-                  stringToParsedValue(
-                      internalRowSchema.getField(valueEntry.getKey().toLowerCase()).getType(),
-                      valueEntry.getValue()));
-        }
 
-        for (Map.Entry<String, String> pkEntry : keyValues.entrySet()) {
-          if (pkEntry.getValue() == null) {
-            continue;
+          for (Map.Entry<String, String> pkEntry : keyValues.entrySet()) {
+            if (pkEntry.getValue() == null) {
+              continue;
+            }
+            // TODO(pabloem): Understand why SpannerSchema has field names in lowercase...
+            rowBuilder =
+                rowBuilder.withFieldValue(
+                    pkEntry.getKey().toLowerCase(),
+                    stringToParsedValue(
+                        internalRowSchema.getField(pkEntry.getKey().toLowerCase()).getType(),
+                        pkEntry.getValue()));
           }
-          // TODO(pabloem): Understand why SpannerSchema has field names in lowercase...
-          rowBuilder =
-              rowBuilder.withFieldValue(
-                  pkEntry.getKey().toLowerCase(),
-                  stringToParsedValue(
-                      internalRowSchema.getField(pkEntry.getKey().toLowerCase()).getType(),
-                      pkEntry.getValue()));
+          receiver
+              .get(OUTPUT_TAG)
+              .outputWithTimestamp(
+                  Row.withSchema(tableChangeRecordSchema)
+                      .addValue(record.getModType().toString())
+                      .addValue(record.getCommitTimestamp().toString())
+                      .addValue(Long.parseLong(record.getRecordSequence()))
+                      .addValue(rowBuilder.build())
+                      .build(),
+                  timestamp);
+        } catch (Exception e) {
+          errorsInBundle += 1;
```

Comment:
Is is it possible for anything here to have an exception? I'm especially looking at .getKeysJson and .getNewValuesJson. If so, we would fail to output to the error queue

Reply:
I discovered that it is possible for the spanner schema to change while this transform is running. It is a case which is handled in the SpannerIO successfully, but our transform doesn't handle it. That is the only mode of failure that I could come up with. As for the .getKeysJson and .getNewValuesJson they are returning the spanner row as JSON from the Mod and mapping it to a hash map. The Mod is created by the I/O itself, so I don't think there will be any formatting issues there and converting them to a hash map should be straightforward.

Reply:
Is it at all possible for there to be duplicate keys?

Reply:
I am not sure. Is it allowed to have two fields with the same name in spanner? If yes then there can be, other possibility is if the SpannerIO messes up.

Reply:
tyvm. I agree its low, but poorly handle exceptions are the causes of recent outages

Reply:
Done.